tuf/client/updater.py

```diff
@@ -650,89 +650,73 @@ def refresh(self, unsafely_update_root_if_necessary=True):
-
-      else:
-        raise
-
-    except tuf.ExpiredMetadataError:
-      if unsafely_update_root_if_necessary:
-        logger.info('No changes were detected from the mirrors for a given role'
-          ', and that metadata that is available on disk has been found to be'
-          ' expired. Trying to update root in case of foul play.')
-        self._update_metadata('root', DEFAULT_ROOT_UPPERLENGTH)
-        self.refresh(unsafely_update_root_if_necessary=False)
-
-      # The caller explicitly requested not to unsafely fetch an expired Root.
-      else:
-        logger.info('No changes were detected from the mirrors for a given role'
-          ', and that metadata that is available on disk has been found to be '
-          'expired. Your metadata is out of date.')
-        raise
-
-
+    self._update_metadata('timestamp', DEFAULT_TIMESTAMP_UPPERLENGTH)
+    self._update_metadata_if_changed('snapshot',
+                                     referenced_metadata='timestamp')
+    self._update_metadata_if_changed('targets')
+    
 
+  def _update_root_metadata(self, current_root_metadata, compression_algorithm=None):
+    """
+      <Purpose>
+        The root file must be signed by the current root threshold and keys as well
+        as the previous root threshold and keys. The update process for root files 
+        means that each intermediate root file must be downloaded, to build a chain 
+        of trusted root keys from keys already trusted by the client:
+        
+          1.root -> 2.root -> 3.root
+         
+        3.root must be signed by the threshold and keys of 2.root, and 2.root must
+        be signed by the threshold and keys of 1.root.
+    
+      <Arguments>
+        current_root_metadata:
+          The currently held version of metadata.
+           
+        compresison_algorithm:
+          The compression algorithm used to compress remote files.
+    
+      <Side Effects>
+        Updates the root metadata files with the latest information.
+    
+      <Returns>
+        None.
+    """
+    next_version = current_root_metadata['version'] + 1
+    new_root_metadata_file = self._get_metadata_file('root', 'root.json', tuf.conf.DEFAULT_ROOT_REQUIRED_LENGTH, None, 
+                                                     compression_algorithm=compression_algorithm)
+    new_root_metadata = \
+      tuf.util.load_json_string(new_root_metadata_file.read().decode('utf-8'))
+    new_version = new_root_metadata['signed']['version'] + 1
+    
+    for version in range(next_version, new_version):
+      # Temporarily set consistent snapshot. Will be updated to whatever is set in the latests root.json after 
+      # running through the intermediates with `_update_metadata`
+      self.consistent_snapshot = True
+      self._update_metadata('root', tuf.conf.DEFAULT_ROOT_REQUIRED_LENGTH, version=version, 
+                            compression_algorithm=compression_algorithm)
```

Comment:
> I am not sure that root.json is properly updated by snapshot.json

It doesn't update the root again if the conditions I listed hold, it simply fails the update process (the hash in snapshot will not be correct).